File: src/main/java/util/ListUtils.java
package util;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class ListUtils {
    @SafeVarargs
    public static <T> List<T> listOf(T... elems) {
        return Stream.of(elems).collect(Collectors.toList());
    }

    @SafeVarargs
    public static <T> List<T> concat(List<T>... lists) {
        return Stream.of(lists)
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }

    public static <T> Optional<T> first(List<T> list) {
        if (list.size() == 0)
            return Optional.empty();
        else
            return Optional.ofNullable(list.get(0));
    }

    public static <T> Optional<T> last(List<T> list) {
        if (list.size() == 0)
            return Optional.empty();
        else
            return Optional.ofNullable(list.get(list.size() - 1));
    }
}

File: src/main/kotlin/org/polystat/j2eo/util/ListUtils.kt
package org.polystat.j2eo.util

import java.util.Optional
import java.util.stream.Collectors
import java.util.stream.Stream

object ListUtils {
    @JvmStatic
    @SafeVarargs
    fun <T> listOf(vararg elems: T): List<T> = Stream.of(*elems).collect(Collectors.toList())

    @SafeVarargs
    fun <T> concat(vararg lists: List<T>): List<T> =
        Stream
            .of(*lists)
            .flatMap { obj: List<T> -> obj.stream() }
            .collect(Collectors.toList())

    fun <T : Any> first(list: List<T>): Optional<T> =
        if (list.isEmpty()) {
            Optional.empty()
        } else {
            Optional.ofNullable(list[0])
        }

    fun <T : Any> last(list: List<T>): Optional<T> =
        if (list.isEmpty()) {
            Optional.empty()
        } else {
            Optional.ofNullable(list[list.size - 1])
        }
}
